File: examples/3d_rotation_conversion.py
import numpy as np
from scipy.spatial.transform import Rotation

# The given 3D rotation
euler = (45, 30, 60) # Unit: [deg] in the XYZ-order

# Generate 3D rotation object
robj = Rotation.from_euler('zyx', euler[::-1], degrees=True)

# Print other representations
print('\n## Euler Angle (ZYX)')
print(np.rad2deg(robj.as_euler('zyx'))) # [60, 30, 45] [deg] in the ZYX-order
print('\n## Rotation Matrix')
print(robj.as_matrix())
print('\n## Rotation Vector')
print(robj.as_rotvec())                 # [0.97, 0.05, 1.17]
print('\n## Quaternion (XYZW)')
print(robj.as_quat())                   # [0.44, 0.02, 0.53, 0.72]

File: examples/3d_rotation_conversion.cpp
#include <iostream>
#include "Eigen/Dense"

int main() {
    // The given 3D rotation
    Eigen::Vector3d euler_degrees(45, 30, 60);  // Unit: [deg] in the XYZ-order (column vector)
    Eigen::Vector3d euler_radians = euler_degrees * M_PI / 180.0; // Unit: radian

    // Rotation matrix converted from euler angles
    Eigen::Matrix3d rotation_matrix = Eigen::AngleAxisd(euler_radians.x(), Eigen::Vector3d::UnitX())
                                      * Eigen::AngleAxisd(euler_radians.y(), Eigen::Vector3d::UnitY())
                                      * Eigen::AngleAxisd(euler_radians.z(), Eigen::Vector3d::UnitZ()).toRotationMatrix();

    // Rotation vector (rodrigues' rotation)
    Eigen::AngleAxisd rotation_vector(rotation_matrix);

    // Quaternion
    Eigen::Quaterniond quaternion(rotation_vector);

    std::cout << "\n## Euler Angle (XYZ)\n" << euler_degrees.transpose() << std::endl;
    std::cout << "\n## Rotation Matrix\n" << rotation_matrix << std::endl;
    std::cout << "\n## Rotation Vector\n" << rotation_vector.angle() * rotation_vector.axis().transpose() << std::endl;
    std::cout << "\n## Quaternion\n" << quaternion.coeffs().transpose() << std::endl;

    return 0;
}